Transpiling junegunns fzf "Cannot internalize utils.Char"

Hello,

I am trying to transpile junegunn/fzf, a really really good fuzzy-finder written in Golang.

My main-Function looks like this:

```
package main

import (
	"github.com/gopherjs/gopherjs/js"
	"github.com/junegunn/fzf/src/algo"
	"github.com/junegunn/fzf/src/util"
)

func main() {
	js.Global.Set("fzf", map[string]interface{}{
		"FuzzyMatchV2": algo.FuzzyMatchV2,
	})
}
```

Calling `fzf.FuzzyMatchV2(true, false, false, "Hello World", "aW", true, null)` leads to an error:

> ..js:1419 Uncaught Error: runtime error: cannot internalize util.Chars

The util.Chars-Struct looks like this:

```
type Chars struct {
	slice           []byte // or []rune
	inBytes         bool
	trimLengthKnown bool
	trimLength      uint16

	// XXX Piggybacking item index here is a horrible idea. But I'm trying to
	// minimize the memory footprint by not wasting padded spaces.
	Index int32
}
```

If you want to try it, use my fork of fzf, branch "gopherjs". There is a small change, which takes out some unix syscall dependencies, to make the build happen.

> https://github.com/dithmer/fzf/tree/gopherjs

I cannot find out what is leading to this error. Any help is appreciated. 

Thanks in advance.

Hi @dithmer, sorry for getting back to you so late. I've tried to reproduce this issue with the latest version of GopherJS using https://github.com/dithmer/fzf/tree/gopherjs, but I'm getting build errors for `golang.org/x/sys/unix`, which GopherJS doesn't support:

```
../../../golang.org/x/sys/unix/affinity_linux.go:14:35: undeclared name: _NCPUBITS
../../../golang.org/x/sys/unix/affinity_linux.go:17:25: undeclared name: cpuMask
../../../golang.org/x/sys/unix/affinity_linux.go:50:27: undeclared name: cpuMask
../../../golang.org/x/sys/unix/fcntl.go:14:30: undeclared name: SYS_FCNTL
../../../golang.org/x/sys/unix/fcntl.go:31:42: undeclared name: Flock_t
../../../golang.org/x/sys/unix/fdset.go:11:12: undeclared name: FdSet
../../../golang.org/x/sys/unix/fdset.go:16:12: undeclared name: FdSet
../../../golang.org/x/sys/unix/fdset.go:21:12: undeclared name: FdSet
../../../golang.org/x/sys/unix/fdset.go:26:12: undeclared name: FdSet
../../../golang.org/x/sys/unix/ioctl.go:47:47: undeclared name: Termios
../../../golang.org/x/sys/unix/ioctl.go:47:47: too many errors
```

Could you provide a minimal reproduction example using the latest version of GopherJS?

--- a/compiler/prelude/jsmapping.js
+++ b/compiler/prelude/jsmapping.js
@@ -384,6 +384,19 @@ var $internalize = (v, t, recv, seen, makeWrapper) => {
             if (o !== noJsObject) {
                 return o;
             }
+            var n = new t.ptr();
+            for (var i = 0; i < t.fields.length; i++) {
+              var f = t.fields[i];
+      
+              if (!f.exported) {
+                continue;
+              }
+              var jsProp = v[f.name];
+      
+              n[f.prop] = $internalize(jsProp, f.typ, recv, seen, makeWrapper);
+            }
+      
+            return n;
     }
     $throwRuntimeError("cannot internalize " + t.string);
 };

--- a/tests/js_test.go
+++ b/tests/js_test.go
@@ -10,6 +10,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/google/go-cmp/cmp"
 	"github.com/gopherjs/gopherjs/js"
 )
 
@@ -308,11 +309,12 @@ func TestInternalizeStruct(t *testing.T) {
 		Name string
 		Age  int
 	}
-	var a Person
+	var a, expected Person
+	expected = Person{Name: "foo", Age: 952}
 
 	js.Global.Set("return_person", func(p *Person) *Person {
 		if p == nil {
-			t.Fail() // Or t.Error("Received nil pointer")
+			t.Fail()
 			return nil
 		}
 		a = *p
@@ -320,19 +322,17 @@ func TestInternalizeStruct(t *testing.T) {
 	})
 
 	js.Global.Call("eval", "return_person({Name: 'foo', Age: 952})")
-
-	if a.Name != "foo" || a.Age != 952 {
-		t.Fail()
+	if diff := cmp.Diff(a, expected); diff != "" {
+		t.Errorf("Mismatch (-want +got):\n%s", diff)
 	}
 }
-
 func TestInternalizeStructUnexportedFields(t *testing.T) {
 	type Person struct {
 		Name string
 		age  int
 	}
-	var a Person
-
+	var a, expected Person
+	expected = Person{Name: "foo", age: 0}
 	js.Global.Set("return_person", func(p *Person) *Person {
 		a = *p
 		return p
@@ -340,8 +340,14 @@ func TestInternalizeStructUnexportedFields(t *testing.T) {
 
 	js.Global.Call("eval", "return_person({Name: 'foo', age: 952})")
 
-	if a.Name != "foo" || a.age != 0 {
-		t.Fail()
+	// Manually check unexported fields
+	if a.age != expected.age {
+		t.Errorf("Mismatch in age: got %v, want %v", a.age, expected.age)
+	}
+
+	// Check exported fields using cmp.Diff
+	if diff := cmp.Diff(a.Name, expected.Name); diff != "" {
+		t.Errorf("Mismatch in Name (-want +got):\n%s", diff)
 	}
 }
 
@@ -355,7 +361,8 @@ func TestInternalizeStructNested(t *testing.T) {
 		Age  int
 		F    FullName
 	}
-	var a Person
+	var a, expected Person
+	expected = Person{Name: "foo", Age: 952, F: FullName{FirstName: "John", LastName: "Doe"}}
 
 	js.Global.Set("return_person", func(p *Person) *Person {
 		a = *p
@@ -363,9 +370,8 @@ func TestInternalizeStructNested(t *testing.T) {
 	})
 
 	js.Global.Call("eval", "return_person({Name: 'foo', Age: 952, F: {FirstName: 'John', LastName: 'Doe'}})")
-
-	if a.Name != "foo" || a.Age != 952 || a.F.FirstName != "John" || a.F.LastName != "Doe" {
-		t.Fail()
+	if diff := cmp.Diff(a, expected); diff != "" {
+		t.Errorf("Mismatch (-want +got):\n%s", diff)
 	}
 }
 
@@ -377,8 +383,8 @@ func TestInternalizeArrayOfStructs(t *testing.T) {
 	type ArrayOfStructs struct {
 		People []Person
 	}
-
-	var a ArrayOfStructs
+	var a, expected ArrayOfStructs
+	expected = ArrayOfStructs{People: []Person{{Name: "Alice", Age: 30}, {Name: "Bob", Age: 40}}}
 
 	js.Global.Set("return_people_array", func(p ArrayOfStructs) ArrayOfStructs {
 		a = p
@@ -386,9 +392,8 @@ func TestInternalizeArrayOfStructs(t *testing.T) {
 	})
 
 	js.Global.Call("eval", `return_people_array({People: [{Name: "Alice", Age: 30}, {Name: "Bob", Age: 40}]})`)
-
-	if len(a.People) != 2 || a.People[0].Name != "Alice" || a.People[1].Age != 40 {
-		t.Fail()
+	if diff := cmp.Diff(a, expected); diff != "" {
+		t.Errorf("Mismatch (-want +got):\n%s", diff)
 	}
 }
 

--- a/tests/js_test.go
+++ b/tests/js_test.go
@@ -303,6 +303,95 @@ func TestInternalizeDate(t *testing.T) {
 	}
 }
 
+func TestInternalizeStruct(t *testing.T) {
+	type Person struct {
+		Name string
+		Age  int
+	}
+	var a Person
+
+	js.Global.Set("return_person", func(p *Person) *Person {
+		if p == nil {
+			t.Fail() // Or t.Error("Received nil pointer")
+			return nil
+		}
+		a = *p
+		return p
+	})
+
+	js.Global.Call("eval", "return_person({Name: 'foo', Age: 952})")
+
+	if a.Name != "foo" || a.Age != 952 {
+		t.Fail()
+	}
+}
+
+func TestInternalizeStructUnexportedFields(t *testing.T) {
+	type Person struct {
+		Name string
+		age  int
+	}
+	var a Person
+
+	js.Global.Set("return_person", func(p *Person) *Person {
+		a = *p
+		return p
+	})
+
+	js.Global.Call("eval", "return_person({Name: 'foo', age: 952})")
+
+	if a.Name != "foo" || a.age != 0 {
+		t.Fail()
+	}
+}
+
+func TestInternalizeStructNested(t *testing.T) {
+	type FullName struct {
+		FirstName string
+		LastName  string
+	}
+	type Person struct {
+		Name string
+		Age  int
+		F    FullName
+	}
+	var a Person
+
+	js.Global.Set("return_person", func(p *Person) *Person {
+		a = *p
+		return p
+	})
+
+	js.Global.Call("eval", "return_person({Name: 'foo', Age: 952, F: {FirstName: 'John', LastName: 'Doe'}})")
+
+	if a.Name != "foo" || a.Age != 952 || a.F.FirstName != "John" || a.F.LastName != "Doe" {
+		t.Fail()
+	}
+}
+
+func TestInternalizeArrayOfStructs(t *testing.T) {
+	type Person struct {
+		Name string
+		Age  int
+	}
+	type ArrayOfStructs struct {
+		People []Person
+	}
+
+	var a ArrayOfStructs
+
+	js.Global.Set("return_people_array", func(p ArrayOfStructs) ArrayOfStructs {
+		a = p
+		return p
+	})
+
+	js.Global.Call("eval", `return_people_array({People: [{Name: "Alice", Age: 30}, {Name: "Bob", Age: 40}]})`)
+
+	if len(a.People) != 2 || a.People[0].Name != "Alice" || a.People[1].Age != 40 {
+		t.Fail()
+	}
+}
+
 func TestEquality(t *testing.T) {
 	if js.Global.Get("Array") != js.Global.Get("Array") || js.Global.Get("Array") == js.Global.Get("String") {
 		t.Fail()

